File: src/main/java/dev/slne/surf/essentials/commands/minecraft/FunctionCommand.java
package dev.slne.surf.essentials.commands.minecraft;

import dev.jorel.commandapi.executors.NativeResultingCommandExecutor;
import dev.jorel.commandapi.wrappers.FunctionWrapper;
import dev.slne.surf.essentials.commands.EssentialsCommand;
import dev.slne.surf.essentials.utils.EssentialsUtil;
import dev.slne.surf.essentials.utils.color.Colors;
import dev.slne.surf.essentials.utils.permission.Permissions;
import lombok.val;
import net.kyori.adventure.text.Component;
import org.bukkit.command.CommandSender;
import org.jetbrains.annotations.NotNull;

public class FunctionCommand extends EssentialsCommand {

    public FunctionCommand() {
        super("function", "function <function>", "Runs a function");

        withPermission(Permissions.FUNCTION_PERMISSION);

        then(functionArgument("function")
                .executesNative((NativeResultingCommandExecutor) (sender, args) -> executeFunction(sender.getCallee(), args.getUnchecked("function"))));
    }

    private int executeFunction(@NotNull CommandSender source, FunctionWrapper @NotNull [] commandFunctions) {
        int executedFunctions = 0;

        for (FunctionWrapper function : commandFunctions) {
            executedFunctions += function.run(source);
        }

        val isSingleFunction = executedFunctions == 1;

        if (commandFunctions.length == 1) {
            EssentialsUtil.sendSuccess(source, Component.text("Es %s ".formatted((isSingleFunction) ? "wurde" : "wurden"), Colors.SUCCESS)
                    .append(Component.text(executedFunctions, Colors.VARIABLE_VALUE))
                    .append(Component.text(" %s von der Funktion ".formatted(isSingleFunction ? "command" : "commands"), Colors.SUCCESS))
                    .append(Component.text(commandFunctions[0].getKey().asString(), Colors.VARIABLE_KEY))
                    .append(Component.text(" ausgeführt.", Colors.SUCCESS)));

        } else {
            EssentialsUtil.sendSuccess(source, Component.text("Es %s ".formatted((isSingleFunction) ? "wurde" : "wurden"), Colors.SUCCESS)
                    .append(Component.text(executedFunctions, Colors.VARIABLE_VALUE))
                    .append(Component.text(" %s aus ".formatted(isSingleFunction ? "command" : "commands"), Colors.SUCCESS))
                    .append(Component.text(commandFunctions.length, Colors.VARIABLE_VALUE))
                    .append(Component.text(" Funktionen ausgeführt.", Colors.SUCCESS)));
        }

        return executedFunctions;
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/command/minecraft/FunctionCommand.kt
package dev.slne.surf.essentials.command.minecraft

import dev.jorel.commandapi.kotlindsl.*
import dev.jorel.commandapi.wrappers.FunctionWrapper
import dev.slne.surf.api.core.messages.adventure.sendText
import dev.slne.surf.essentials.util.permission.EssentialsPermissionRegistry
import dev.slne.surf.essentials.util.util.isFolia
import org.bukkit.Bukkit
import org.bukkit.entity.Entity

fun functionCommand() = commandTree("function") {
    withPermission(EssentialsPermissionRegistry.FUNCTION_COMMAND)
    functionArgument("function") {
        anyExecutor { executor, args ->
            if (Bukkit.getServer().isFolia()) {
                executor.sendText {
                    appendErrorPrefix()
                    error("Der Function-Befehl wird auf Folia-Servern nicht unterstützt.")
                }
                return@anyExecutor
            }

            val function: FunctionWrapper by args
            function.run()

            executor.sendText {
                appendSuccessPrefix()
                success("Die Funktion ")
                variableValue(function.key.toString())
                success(" wurde ausgeführt.")
            }
        }

        literalArgument("as") {
            entitySelectorArgumentOneEntity("entity") {
                anyExecutor { executor, args ->
                    if (Bukkit.getServer().isFolia()) {
                        executor.sendText {
                            appendErrorPrefix()
                            error("Der Function-Befehl wird auf Folia-Servern nicht unterstützt.")
                        }
                        return@anyExecutor
                    }

                    val function: FunctionWrapper by args
                    val entity: Entity by args

                    function.runAs(entity)

                    executor.sendText {
                        appendSuccessPrefix()
                        success("Die Funktion ")
                        variableValue(function.key.toString())
                        success(" wurde als ")
                        variableValue(entity.name)
                        success(" ausgeführt.")
                    }
                }
            }
        }
    }
}